### PR TITLE
#162948907 generate and return jwt tokens

### PR DESCRIPTION
**Pull request by danuluma**
#### What does this PR do?
Users should receive a JWT on successful Registration 

#### Description of Task(s) to be completed?
return a JWT to a user on successful registration
generate and return a JWT when a user logs in

#### How should this be manually tested?
* Pull the branch `git fetch origin ft-return-token-162948907:ft-return-token-162948907`
* Checkout to the branch `git checkout ft-return-token-162948907`
* Install dependencies `pip3 install -r requirements.txt`
* Run the server `python3 manage.py runserver`
* Use postman to send a `POST` request to `http://127.0.0.1:8000/api/users/`
* The data should be in the format of:
```
{   "user":{
    "username": "dann",
    "email": "dann@dan.com",
    "password": "county001"
    }
 }
```
* Remember to select `application/json` as the data type
* The response should be a json object containing the user's details as well as the jwt token
```
{
    "user": {
        "email": "dann@dan.com",
        "username": "dann",
        "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoxNiwidXNlcm5hbWUiOiJkYW5uQGRhbi5jb20iLCJleHAiOjE1NDc1NzgyMDgsImVtYWlsIjoiZGFubkBkYW4uY29tIn0.n8mW-ncEXc37EwALIlQgF6CVyMtdjKWuowy6SYjamPA"
    }
}
```

#### Any background context you want to provide?
* This project uses the latest version of python3. Please download it from [here](https://www.python.org/downloads/)
* You may as well use pip instead of pip3 if you have either only python3 installed on your computer or you have it set as the default version of python
* Postman can be downloaded from [here](https://www.getpostman.com/downloads/)
* This token needs to be sent in the headers of subsequent requests to protected endpoints
* The header prefix expected is `Bearer`. On postman, this would hence be sent as a Bearer Token
* I also downgraded django-rest-swagger from v2.2.0 to v2.1.2 due to a bug where the authorization headers aren't passed in the request. It is well described [here](https://github.com/marcgibbons/django-rest-swagger/issues/762).

#### What are the relevant pivotal tracker stories?
[#162948907](https://www.pivotaltracker.com/story/show/162948907)

### CODE REVIEW

**Review by Ruiru11 — COMMENTED** [review]
We agreed we should have one commit

**Review by boltonotieno — APPROVED** [review]

**Comment by danuluma** [comment]
I'm fixing that

**Review by Deekerubo — APPROVED** [review]

**Review by Ruiru11 — APPROVED** [review]

File: authors/apps/authentication/backends.py [hunk 1]
```diff
@@ -1,10 +1,31 @@
...
+# rest_framework_jwt
+
+from rest_framework_jwt.settings import api_settings
+from rest_framework_jwt.utils import jwt_get_secret_key
+
 
 """Configure JWT Here"""
 
+
+class JWTokens(object):
```
Comment (tesh254) [inline authors/apps/authentication/backends.py:7]: This class should have a docstring as a PEP8 standard

File: authors/apps/authentication/models.py [hunk 2]
```diff
@@ -4,117 +4,122 @@
 
 from django.conf import settings
 from django.contrib.auth.models import (
-    AbstractBaseUser, BaseUserManager, PermissionsMixin
+		AbstractBaseUser, BaseUserManager, PermissionsMixin
```
Comment (tesh254) [inline authors/apps/authentication/models.py:7]: Also this space right please fix them

File: authors/apps/authentication/models.py [hunk 3]
```diff
@@ -4,117 +4,122 @@
...
+    user = self.create_user(username, email, password)
+    user.is_superuser = True
+    user.is_staff = True
+    user.save()
 
-      return user
+    return user
 
 
 class User(AbstractBaseUser, PermissionsMixin):
```
Comment (tesh254) [inline authors/apps/authentication/models.py:59]: Lacks a docstring to describe the class

File: authors/apps/authentication/serializers.py [hunk 4]
```diff
@@ -23,7 +52,48 @@ class Meta:
...
+        user_qs = User.objects.filter(email=email)
+        if user_qs.exists():
+            raise ValidationError("Email has already been used")
+        elif re.search(r"^[\w\.\+\-]+\@[\w]+\.[a-z]{2,3}$", email) is None:
+            raise ValidationError("Please input a valid email.")
+        return value
+
+    def validate_username(self, value):
+        #get initial state of data as an object and pick the value of username
+        data = self.get_initial()
```
Comment (Ruiru11) [inline authors/apps/authentication/serializers.py]: we agreed we are getting rid of this

File: authors/apps/authentication/serializers.py [hunk 5]
```diff
@@ -34,7 +104,7 @@ class LoginSerializer(serializers.Serializer):
     email = serializers.CharField(max_length=255)
     username = serializers.CharField(max_length=255, read_only=True)
     password = serializers.CharField(max_length=128, write_only=True)
-
+    token = serializers.CharField(max_length=128, read_only=True)
 
     def validate(self, data):
         # The `validate` method is where we make sure that the current
```
Comment (Ruiru11) [inline authors/apps/authentication/serializers.py:78]: This comment is incomplete

File: authors/apps/authentication/backends.py (hunk 1 above)
Comment (danuluma) [reply authors/apps/authentication/backends.py:7]: Thanks,,,,correcting that

File: authors/apps/authentication/models.py (hunk 2 above)
Comment (danuluma) [reply authors/apps/authentication/models.py:7]: OK

File: authors/apps/authentication/models.py (hunk 3 above)
Comment (danuluma) [reply authors/apps/authentication/models.py:59]: Please do check again

File: authors/apps/authentication/serializers.py (hunk 5 above)
Comment (danuluma) [reply authors/apps/authentication/serializers.py:78]: Not quite, please re-check

File: authors/apps/authentication/models.py [hunk 6]
```diff
@@ -6,115 +6,120 @@
 from django.contrib.auth.models import (
 		AbstractBaseUser, BaseUserManager, PermissionsMixin
 )
-from django.db import models 
+from django.db import models
+# from rest_framework_jwt.settings import api_settings
```
Comment (frankip) [inline authors/apps/authentication/models.py]: @danuluma  Remove commented out code

Comment (danuluma) [reply authors/apps/authentication/models.py]: I've done that, thanks

**Review by frankip — APPROVED** [review]
LGTM

**Pull request merged**